File: idea/src/org/jetbrains/jet/plugin/quickfix/ChangeVariableMutabilityFix.java
/*
 * Copyright 2010-2013 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.jet.plugin.quickfix;

import com.intellij.codeInsight.intention.IntentionAction;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.IncorrectOperationException;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.jet.lang.descriptors.VariableDescriptor;
import org.jetbrains.jet.lang.psi.JetFile;
import org.jetbrains.jet.lang.psi.JetProperty;
import org.jetbrains.jet.lang.psi.JetSimpleNameExpression;
import org.jetbrains.jet.lang.resolve.BindingContext;
import org.jetbrains.jet.lang.resolve.BindingContextUtils;
import org.jetbrains.jet.lang.resolve.DescriptorToSourceUtils;
import org.jetbrains.jet.plugin.JetBundle;
import org.jetbrains.jet.plugin.caches.resolve.ResolvePackage;

import static org.jetbrains.jet.lang.psi.PsiPackage.JetPsiFactory;

public class ChangeVariableMutabilityFix implements IntentionAction {
    private boolean isVar;

    public ChangeVariableMutabilityFix(boolean isVar) {
        this.isVar = isVar;
    }

    public ChangeVariableMutabilityFix() {
        this(false);
    }

    @NotNull
    @Override
    public String getText() {
        return isVar ? JetBundle.message("make.variable.immutable") : JetBundle.message("make.variable.mutable");
    }

    @NotNull
    @Override
    public String getFamilyName() {
        return JetBundle.message("change.variable.mutability.family");
    }

    @Override
    public boolean isAvailable(@NotNull Project project, Editor editor, PsiFile file) {
        if (!(file instanceof JetFile)) return false;
        JetProperty property = getCorrespondingProperty(editor, (JetFile) file);
        return property != null && !property.isVar();
    }

    private static JetProperty getCorrespondingProperty(Editor editor, JetFile file) {
        PsiElement elementAtCaret = file.findElementAt(editor.getCaretModel().getOffset());
        JetProperty property = PsiTreeUtil.getParentOfType(elementAtCaret, JetProperty.class);
        if (property != null) return property;
        JetSimpleNameExpression simpleNameExpression = PsiTreeUtil.getParentOfType(elementAtCaret, JetSimpleNameExpression.class);
        if (simpleNameExpression != null) {
            BindingContext bindingContext = ResolvePackage.getBindingContext(file);
            VariableDescriptor descriptor = BindingContextUtils.extractVariableDescriptorIfAny(bindingContext, simpleNameExpression, true);
            if (descriptor != null) {
                PsiElement declaration = DescriptorToSourceUtils.descriptorToDeclaration(descriptor);
                if (declaration instanceof JetProperty) {
                    return (JetProperty) declaration;
                }
            }
        }
        return null;
    }

    @Override
    public void invoke(@NotNull Project project, Editor editor, PsiFile file) throws IncorrectOperationException {
        JetProperty property = getCorrespondingProperty(editor, (JetFile)file);
        assert property != null;
        property.getValOrVarNode().getPsi().replace(JetPsiFactory(property).createValOrVarNode(isVar ? "val" : "var").getPsi());
    }

    @Override
    public boolean startInWriteAction() {
        return true;
    }
}

File: idea/src/org/jetbrains/kotlin/idea/quickfix/ChangeVariableMutabilityFix.kt
/*
 * Copyright 2010-2015 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.kotlin.idea.quickfix

import com.intellij.codeInsight.intention.IntentionAction
import org.jetbrains.kotlin.descriptors.DeclarationDescriptor
import org.jetbrains.kotlin.diagnostics.Diagnostic
import org.jetbrains.kotlin.diagnostics.DiagnosticFactory1
import org.jetbrains.kotlin.diagnostics.Errors
import org.jetbrains.kotlin.idea.KotlinBundle
import org.jetbrains.kotlin.idea.quickfix.createFromUsage.createCallable.CreatePropertyDelegateAccessorsActionFactory
import org.jetbrains.kotlin.lexer.KtTokens
import org.jetbrains.kotlin.psi.KtProperty
import org.jetbrains.kotlin.psi.KtValVarKeywordOwner
import org.jetbrains.kotlin.psi.psiUtil.getStrictParentOfType
import org.jetbrains.kotlin.resolve.DescriptorToSourceUtils
import org.jetbrains.kotlin.util.OperatorNameConventions

class ReassignmentActionFactory(val factory: DiagnosticFactory1<*, DeclarationDescriptor>) : KotlinSingleIntentionActionFactory() {
    override fun createAction(diagnostic: Diagnostic): IntentionAction? {
        val propertyDescriptor = factory.cast(diagnostic).a
        val declaration =
            DescriptorToSourceUtils.descriptorToDeclaration(propertyDescriptor) as? KtValVarKeywordOwner ?: return null
        return ChangeVariableMutabilityFix(declaration, true)
    }
}

// TODO: Move this to idea-fir-independent
object ConstValFactory : KotlinSingleIntentionActionFactory() {
    override fun createAction(diagnostic: Diagnostic): IntentionAction? {
        val (modifier, element) = Errors.WRONG_MODIFIER_TARGET.cast(diagnostic).run { a to psiElement }
        if (modifier != KtTokens.CONST_KEYWORD) return null
        val property = element.getStrictParentOfType<KtProperty>() ?: return null
        return ChangeVariableMutabilityFix(property, makeVar = false)
    }
}

object DelegatedPropertyValFactory : KotlinSingleIntentionActionFactory() {
    override fun createAction(diagnostic: Diagnostic): IntentionAction? {
        val element = Errors.DELEGATE_SPECIAL_FUNCTION_MISSING.cast(diagnostic).psiElement
        val property = element.getStrictParentOfType<KtProperty>() ?: return null
        val info = CreatePropertyDelegateAccessorsActionFactory.extractFixData(property, diagnostic).singleOrNull() ?: return null
        if (info.name != OperatorNameConventions.SET_VALUE.asString()) return null
        return ChangeVariableMutabilityFix(property, makeVar = false, actionText = KotlinBundle.message("change.to.val"))
    }
}